larq_compute_engine/tflite/kernels/bconv2d.cc

```diff
@@ -199,6 +170,35 @@ TfLiteStatus Prepare(KernelType kernel_type,
   TF_LITE_ENSURE_EQ(context, NumDimensions(post_activation_multiplier), 1);
   TF_LITE_ENSURE_EQ(context, NumDimensions(post_activation_bias), 1);
 
+  conv_params->read_bitpacked_input = input->type == kTfLiteInt32;
+  conv_params->write_bitpacked_output = output->type == kTfLiteInt32;
+
+  if (conv_params->padding_type == TfLitePadding::kTfLitePaddingSame &&
+      conv_params->pad_value != 1 && conv_params->write_bitpacked_output) {
+    context->ReportError(context,
+                         "Writing bitpacked output is only supported with "
+                         "valid or one-padding.");
+    return kTfLiteError;
+  }
+
+  // Read the input dimensions. TF Lite has the same input format as TensorFlow:
+  // (B, H, W, Ci).
+  conv_params->batch = input->dims->data[0];
+  conv_params->input_height = input->dims->data[1];
+  conv_params->input_width = input->dims->data[2];
+  if (!conv_params->read_bitpacked_input) {
+    conv_params->channels_in = input->dims->data[3];
```

Comment:
I would opt for always assuming `channels_in` is set now, and giving an error if it isn't. This will also make sure that "the kernel version matches the converter version" so to speak. I guess if `channels_in` isn't set then we could also throw an error in `Init` already.
```suggestion
    TF_LITE_ENSURE_EQ(context, conv_params->channels_in, input->dims->data[3]);
```

Reply:
Out of interest, do we have a policy on which kernel versions support which converter versions?

Reply:
I guess not. It might be a good idea to start doing that. I think we are planning to change the attributes to be more in line with the tflite conv, (i.e. strings -> enums). When we have that, maybe we could label that as "version 1" and then use tflite's op versioning system whenever we have a breaking change from that point.